File: backend-dotnet/src/Lmp.Infrastructure/SiteConfiguration/SiteConfigManager.cs
using System.Collections.Concurrent;
using System.Text.Json;
using Lmp.Application.SiteConfiguration;
using Lmp.Domain.SiteConfiguration;
using Lmp.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace Lmp.Infrastructure.SiteConfiguration;

/// <summary>
/// Central site-configuration manager. Faithful port of
/// <c>com.lmp.shared.config.site.SiteConfigManager</c>: a hierarchical resolver
/// (env/config → <c>site-config.json</c> → <c>site_config</c> table → values
/// derived from <c>lmp.site.url</c>) behind a memory cache. Registered as a
/// singleton; database reads use a scoped <see cref="LmpDbContext"/>.
/// </summary>
public sealed class SiteConfigManager : ISiteConfigManager
{
    private readonly IConfiguration _configuration;
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly ILogger<SiteConfigManager> _logger;

    private readonly ConcurrentDictionary<string, string?> _cache = new(StringComparer.Ordinal);
    private readonly IReadOnlyDictionary<string, string> _fileConfig;
    private readonly IReadOnlyDictionary<string, string> _derivedConfig;
    private readonly string _siteUrl;

    public SiteConfigManager(
        IConfiguration configuration,
        IServiceScopeFactory scopeFactory,
        IOptions<SiteConfigOptions> options,
        ILogger<SiteConfigManager> logger)
    {
        _configuration = configuration;
        _scopeFactory = scopeFactory;
        _logger = logger;

        var url = options.Value.Url ?? "http://localhost:8080";
        _siteUrl = url.TrimEnd('/');
        _derivedConfig = ComputeDerivedValues(_siteUrl);
        _fileConfig = LoadFileConfig(options.Value.ConfigPath);
        _logger.LogInformation("SiteConfigManager initialised - siteUrl={SiteUrl}", _siteUrl);
    }

    public string? GetString(string key) => _cache.GetOrAdd(key, Resolve);

    public string GetString(string key, string defaultValue) => GetString(key) ?? defaultValue;

    public bool GetBoolean(string key, bool defaultValue)
    {
        try
        {
            var value = GetString(key);
            return value is null ? defaultValue : bool.TryParse(value.Trim(), out var b) && b;
        }
        catch (Exception e)
        {
            _logger.LogWarning(e, "[SITE-CONFIG] getBoolean({Key}) failed - fallback {Default}", key, defaultValue);
            return defaultValue;
        }
    }

    public string GetBaseUrl() => GetString("app.base.url", _siteUrl);

    public string GetFrontendUrl() => GetString("app.frontend.url", GetBaseUrl());

    public string GetSiteName() => GetString("app.name", "LMP Digital Services");

    public string GetSupportEmail() => GetString("mail.from.support", "support@" + ExtractHost(GetBaseUrl()));

    public string GetContactEmail() => GetString("mail.from.contact", "info@" + ExtractHost(GetBaseUrl()));

    public string GetNoreplyEmail() => GetString("mail.from.noreply", "noreply@" + ExtractHost(GetBaseUrl()));

    public string GetCompanyWebsite() => GetString("company.website", GetBaseUrl());

    public string GetOauth2IssuerUri() => GetString("app.oauth2.issuer-uri", GetBaseUrl());

    public string GetCorsAllowedOrigins() => GetString("app.cors.allowed-origins", GetBaseUrl());

    public async Task UpdateAsync(string key, string value, string? description, CancellationToken ct = default)
    {
        await using var scope = _scopeFactory.CreateAsyncScope();
        var db = scope.ServiceProvider.GetRequiredService<LmpDbContext>();

        var entry = await db.SiteConfig.FindAsync([key], ct);
        if (entry is null)
        {
            entry = new SiteConfigEntry(key, value, description);
            db.SiteConfig.Add(entry);
        }
        else
        {
            entry.Value = value;
            if (description is not null)
            {
                entry.Description = description;
            }
        }

        entry.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);

        _cache.TryRemove(key, out _);
        _logger.LogInformation("SiteConfig updated - {Key} = {Value}", key, value);
    }

    public async Task<IReadOnlyDictionary<string, string?>> GetAllFromDbAsync(CancellationToken ct = default)
    {
        await using var scope = _scopeFactory.CreateAsyncScope();
        var db = scope.ServiceProvider.GetRequiredService<LmpDbContext>();
        var entries = await db.SiteConfig.AsNoTracking().OrderBy(e => e.Key).ToListAsync(ct);
        return entries.ToDictionary(e => e.Key, e => e.Value, StringComparer.Ordinal);
    }

    private string? Resolve(string key)
    {
        // 1. Environment / configuration providers.
        var envValue = _configuration[key];
        if (!string.IsNullOrWhiteSpace(envValue))
        {
            return envValue;
        }

        // 2. site-config.json.
        if (_fileConfig.TryGetValue(key, out var fileValue))
        {
            return fileValue;
        }

        // 3. Database — fail-soft: a not-yet-ready DB must not crash resolution.
        try
        {
            using var scope = _scopeFactory.CreateScope();
            var db = scope.ServiceProvider.GetRequiredService<LmpDbContext>();
            var dbValue = db.SiteConfig.AsNoTracking()
                .Where(e => e.Key == key)
                .Select(e => e.Value)
                .FirstOrDefault();
            if (dbValue is not null)
            {
                return dbValue;
            }
        }
        catch (Exception e)
        {
            _logger.LogWarning("[SITE-CONFIG] DB read unavailable for '{Key}' - derived/env fallback: {Message}", key, e.Message);
        }

        // 4. Values derived from lmp.site.url.
        return _derivedConfig.GetValueOrDefault(key);
    }

    private static Dictionary<string, string> ComputeDerivedValues(string siteUrl)
    {
        var derived = new Dictionary<string, string>(StringComparer.Ordinal);
        if (string.IsNullOrWhiteSpace(siteUrl))
        {
            return derived;
        }

        var host = ExtractHost(siteUrl);
        var isLocal = host is "localhost" or "127.0.0.1";
        var hostNoWww = host.StartsWith("www.", StringComparison.Ordinal) ? host[4..] : host;

        derived["app.base.url"] = siteUrl;
        derived["app.frontend.url"] = siteUrl;
        derived["company.website"] = siteUrl;
        // Single-host monolith: issuer = site URL (auth pages on the same host).
        derived["app.oauth2.issuer-uri"] = siteUrl;
        derived["app.cors.allowed-origins"] = isLocal
            ? "http://localhost:*"
            : $"https://{hostNoWww},https://www.{hostNoWww}";

        derived["mail.from.noreply"] = "noreply@" + host;
        derived["mail.from.support"] = "support@" + host;
        derived["mail.replyto.support"] = "support@" + host;
        derived["company.email"] = "support@" + host;
        derived["company.team.email"] = "support@" + host;
        derived["company.admin.email"] = "admin@" + host;
        derived["lmp.sync.alert.admin-email"] = "admin@" + host;

        return derived;
    }

    private Dictionary<string, string> LoadFileConfig(string? configPath)
    {
        var path = !string.IsNullOrWhiteSpace(configPath)
            ? configPath
            : Path.Combine(AppContext.BaseDirectory, "site-config.json");

        if (!File.Exists(path))
        {
            return new Dictionary<string, string>(StringComparer.Ordinal);
        }

        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            var flattened = new Dictionary<string, string>(StringComparer.Ordinal);
            Flatten(doc.RootElement, string.Empty, flattened);
            return flattened;
        }
        catch (Exception e)
        {
            _logger.LogWarning("Unable to load site-config.json: {Message}", e.Message);
            return new Dictionary<string, string>(StringComparer.Ordinal);
        }
    }

    private static void Flatten(JsonElement element, string prefix, Dictionary<string, string> target)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var property in element.EnumerateObject())
            {
                var key = prefix.Length == 0 ? property.Name : $"{prefix}.{property.Name}";
                Flatten(property.Value, key, target);
            }
        }
        else
        {
            target[prefix] = element.ToString();
        }
    }

    private static string ExtractHost(string url)
    {
        if (Uri.TryCreate(url, UriKind.Absolute, out var uri) && !string.IsNullOrEmpty(uri.Host))
        {
            return uri.Host;
        }

        var stripped = url.Replace("https://", string.Empty, StringComparison.OrdinalIgnoreCase)
            .Replace("http://", string.Empty, StringComparison.OrdinalIgnoreCase);
        var slash = stripped.IndexOf('/', StringComparison.Ordinal);
        if (slash > 0)
        {
            stripped = stripped[..slash];
        }

        var colon = stripped.IndexOf(':', StringComparison.Ordinal);
        if (colon > 0)
        {
            stripped = stripped[..colon];
        }

        return stripped;
    }
}

File: src/main/java/com/lmp/shared/config/site/SiteConfigManager.java
package com.lmp.shared.config.site;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.core.env.Environment;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.net.URI;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Gestionnaire central de configuration du site.
 *
 * <p>Sources (par ordre de priorité) :</p>
 * <ol>
 *   <li>Variables d'environnement / properties Spring (ex: SITE_URL)</li>
 *   <li>Fichier {@code site-config.json} (reloadable à chaud)</li>
 *   <li>Base de données {@code site_config} (modifiable via admin)</li>
 *   <li>Valeurs dérivées de {@code lmp.site.url}</li>
 * </ol>
 *
 * <p>Toutes les lectures passent par un cache mémoire.
 * Le cache est invalidé lors d'un reload ou d'une mise à jour DB.</p>
 */
@Component
public class SiteConfigManager {

    private static final Logger logger = LoggerFactory.getLogger(SiteConfigManager.class);

    private final SiteConfigRepository repository;
    private final ApplicationEventPublisher eventPublisher;
    private final Environment environment;

    @Value("${lmp.site.url:http://localhost:8080}")
    private String siteUrl;

    @Value("${lmp.site.config.path:}")
    private String externalConfigPath;

    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Map<String, String> cache = new ConcurrentHashMap<>();
    private Map<String, String> fileConfig = Collections.emptyMap();
    private Map<String, String> derivedConfig = Collections.emptyMap();

    public SiteConfigManager(SiteConfigRepository repository,
                             ApplicationEventPublisher eventPublisher,
                             Environment environment) {
        this.repository = repository;
        this.eventPublisher = eventPublisher;
        this.environment = environment;
    }

    @PostConstruct
    public void init() {
        computeDerivedValues();
        reloadFileConfig();
        warmCache();
        logger.info("SiteConfigManager initialisé — siteUrl={}", siteUrl);
    }

    // ============ Lecture publique ============

    public String getString(String key) {
        return cache.computeIfAbsent(key, this::resolve);
    }

    public String getString(String key, String defaultValue) {
        String value = getString(key);
        return value != null ? value : defaultValue;
    }

    /**
     * Lit un booléen de la config hiérarchique (env → file → DB → défaut).
     * Fail-open : valeur absente OU erreur de lecture ⇒ defaultValue —
     * une panne de config ne doit jamais couper silencieusement une intégration.
     * Toute valeur non nulle autre que {@code "true"} (insensible à la casse)
     * est traitée comme {@code false}.
     */
    public boolean getBoolean(String key, boolean defaultValue) {
        try {
            String value = getString(key);
            return value == null ? defaultValue : Boolean.parseBoolean(value.trim());
        } catch (RuntimeException e) {
            logger.warn("[SITE-CONFIG] getBoolean({}) en échec — fallback {}", key, defaultValue, e);
            return defaultValue;
        }
    }

    public String getBaseUrl() {
        return getString("app.base.url", siteUrl);
    }

    public String getFrontendUrl() {
        return getString("app.frontend.url", getBaseUrl());
    }

    public String getSiteName() {
        return getString("app.name", "LMP Digital Services");
    }

    public String getSupportEmail() {
        return getString("mail.from.support", "support@" + extractHost(getBaseUrl()));
    }

    public String getContactEmail() {
        return getString("mail.from.contact", "info@" + extractHost(getBaseUrl()));
    }

    public String getNoreplyEmail() {
        return getString("mail.from.noreply", "noreply@" + extractHost(getBaseUrl()));
    }

    public String getCompanyWebsite() {
        return getString("company.website", getBaseUrl());
    }

    public String getOauth2IssuerUri() {
        return getString("app.oauth2.issuer-uri", getBaseUrl());
    }

    public String getCorsAllowedOrigins() {
        return getString("app.cors.allowed-origins", getBaseUrl());
    }

    // ============ Administration ============

    /**
     * Met à jour une valeur en base de données et invalide le cache.
     */
    public void update(String key, String value, String description) {
        SiteConfigEntry entry = repository.findByKey(key)
            .orElse(new SiteConfigEntry(key, value, description));
        entry.setValue(value);
        if (description != null) {
            entry.setDescription(description);
        }
        repository.save(entry);

        // Invalider le cache et publier l'événement
        String oldValue = cache.get(key);
        cache.remove(key);
        eventPublisher.publishEvent(new SiteConfigChangedEvent(this, key, value));

        logger.info("SiteConfig mise à jour — {} : {} -> {}", key, oldValue, value);
    }

    /**
     * Recharge le fichier {@code site-config.json} et invalide le cache.
     */
    public void reloadFileConfig() {
        Map<String, String> previous = new ConcurrentHashMap<>(fileConfig);
        loadFileConfig();

        // Invalider les clés qui ont changé dans le fichier
        for (String key : fileConfig.keySet()) {
            if (!fileConfig.get(key).equals(previous.get(key))) {
                cache.remove(key);
                eventPublisher.publishEvent(new SiteConfigChangedEvent(this, key, fileConfig.get(key)));
            }
        }
        logger.info("site-config.json rechargé — {} clés", fileConfig.size());
    }

    /**
     * Recharge complètement (fichier + DB) et vide le cache.
     */
    public void reloadAll() {
        cache.clear();
        reloadFileConfig();
        warmCache();
        logger.info("SiteConfigManager reload complet effectué");
    }

    public Map<String, String> getAllFromDb() {
        Map<String, String> result = new ConcurrentHashMap<>();
        repository.findAllByOrderByKeyAsc().forEach(e -> result.put(e.getKey(), e.getValue()));
        return result;
    }

    // ============ Résolution interne ============

    private String resolve(String key) {
        // 1. Variables d'environnement / properties Spring
        String envValue = environment.getProperty(key);
        if (envValue != null && !envValue.isBlank()) {
            return envValue;
        }

        // 2. Fichier site-config.json
        if (fileConfig.containsKey(key)) {
            return fileConfig.get(key);
        }

        // 3. Base de données — fail-soft : DB pas prête au boot (ou blip réseau)
        //    ne doit pas crasher le démarrage, on retombe sur le tier dérivé.
        try {
            Optional<SiteConfigEntry> dbEntry = repository.findByKey(key);
            if (dbEntry.isPresent() && dbEntry.get().getValue() != null) {
                return dbEntry.get().getValue();
            }
        } catch (RuntimeException e) {
            logger.warn("[SITE-CONFIG] Lecture DB indisponible pour '{}' — fallback dérivé/env : {}",
                    key, e.getMessage());
        }

        // 4. Valeurs dérivées de lmp.site.url
        return derivedConfig.get(key);
    }

    private void computeDerivedValues() {
        Map<String, String> derived = new ConcurrentHashMap<>();
        if (siteUrl == null || siteUrl.isBlank()) {
            this.derivedConfig = derived;
            return;
        }
        siteUrl = siteUrl.replaceAll("/+$", "");
        String host = extractHost(siteUrl);
        boolean isLocal = "localhost".equals(host) || "127.0.0.1".equals(host);
        String hostNoWww = host.startsWith("www.") ? host.substring(4) : host;

        derived.put("app.base.url", siteUrl);
        derived.put("app.frontend.url", siteUrl);
        derived.put("company.website", siteUrl);
        // Single-host monolith : issuer = site URL (auth pages servies sur le même host).
        derived.put("app.oauth2.issuer-uri", siteUrl);
        derived.put("app.cors.allowed-origins",
            isLocal
                ? "http://localhost:*"
                : "https://" + hostNoWww + ",https://www." + hostNoWww);

        derived.put("mail.from.noreply", "noreply@" + host);
        derived.put("mail.from.support", "support@" + host);
        derived.put("mail.replyto.support", "support@" + host);
        derived.put("company.email", "support@" + host);
        derived.put("company.team.email", "support@" + host);
        derived.put("company.admin.email", "admin@" + host);
        derived.put("lmp.sync.alert.admin-email", "admin@" + host);

        this.derivedConfig = Collections.unmodifiableMap(derived);
    }

    private void loadFileConfig() {
        Resource resource = resolveConfigResource();
        if (!resource.exists()) {
            this.fileConfig = Collections.emptyMap();
            return;
        }
        try {
            Map<String, Object> raw = objectMapper.readValue(resource.getInputStream(), new TypeReference<>() {});
            Map<String, String> flattened = new ConcurrentHashMap<>();
            flatten(raw, "", flattened);
            this.fileConfig = Collections.unmodifiableMap(flattened);
        } catch (IOException e) {
            logger.warn("Impossible de charger site-config.json : {}", e.getMessage());
            this.fileConfig = Collections.emptyMap();
        }
    }

    private Resource resolveConfigResource() {
        if (externalConfigPath != null && !externalConfigPath.isBlank()) {
            return new FileSystemResource(externalConfigPath);
        }
        return new ClassPathResource("site-config.json");
    }

    private void flatten(Map<String, Object> source, String prefix, Map<String, String> target) {
        for (Map.Entry<String, Object> entry : source.entrySet()) {
            String key = prefix.isEmpty() ? entry.getKey() : prefix + "." + entry.getKey();
            if (entry.getValue() instanceof Map<?, ?> map) {
                @SuppressWarnings("unchecked")
                Map<String, Object> nested = (Map<String, Object>) map;
                flatten(nested, key, target);
            } else {
                target.put(key, String.valueOf(entry.getValue()));
            }
        }
    }

    private void warmCache() {
        // Pré-charger les clés fréquentes
        String[] keys = {
            "app.base.url", "app.frontend.url", "app.name", "app.oauth2.issuer-uri",
            "company.website", "mail.from.support", "mail.from.noreply",
            "company.email", "company.admin.email", "app.cors.allowed-origins"
        };
        for (String key : keys) {
            cache.put(key, resolve(key));
        }
    }

    private String extractHost(String url) {
        try {
            return URI.create(url).getHost();
        } catch (Exception e) {
            String stripped = url.replaceAll("^https?://", "");
            int slashIdx = stripped.indexOf('/');
            if (slashIdx > 0) stripped = stripped.substring(0, slashIdx);
            int portIdx = stripped.indexOf(':');
            if (portIdx > 0) stripped = stripped.substring(0, portIdx);
            return stripped;
        }
    }
}
